File: src/test/java/com/thebluehats/server/game/enchants/SprintDrainTest.java
package com.thebluehats.server.game.enchants;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import com.thebluehats.server.game.managers.combat.templates.ArrowHitPlayerTemplate;
import com.thebluehats.server.game.managers.combat.templates.EventTemplate;
import com.thebluehats.server.game.managers.enchants.CustomEnchant;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.junit.Test;

public class SprintDrainTest {

    @Test
    public void DamagerGetsSpeedWhenArrowHitsPlayer() {
        final int SPEED_DURATION = 3;
        final int SPEED_AMPLIFIER = 1;

        Player player = mock(Player.class);
        Player damagee = mock(Player.class);

        CustomEnchant<SprintDrainArgs> enchant = new SprintDrain(new EventTemplate[] { new ArrowHitPlayerTemplate() });

        enchant.execute(new SprintDrainArgs(player, damagee, SPEED_DURATION, SPEED_AMPLIFIER, 1));
        verify(player).addPotionEffect(new PotionEffect(PotionEffectType.SPEED, SPEED_DURATION * 20, SPEED_AMPLIFIER));
    }

    @Test
    public void DamageeGetsSlownessWhenHitByArrow() {
        final int SPEED_DURATION = 3;
        final int SPEED_AMPLIFIER = 1;

        Player player = mock(Player.class);
        Player damagee = mock(Player.class);

        CustomEnchant<SprintDrainArgs> enchant = new SprintDrain(new EventTemplate[] { new ArrowHitPlayerTemplate() });

        enchant.execute(new SprintDrainArgs(player, damagee, SPEED_DURATION, SPEED_AMPLIFIER, 2));
        verify(damagee).addPotionEffect(new PotionEffect(PotionEffectType.SLOW, 60, 0));
    }
}

File: src/test/kotlin/com/thebluehats/server/game/enchants/SprintDrainTest.kt
package com.thebluehats.server.game.enchants

import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData
import io.kotest.core.spec.style.DescribeSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.bukkit.entity.Player
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class SprintDrainTest : DescribeSpec({
    describe("sprint drain") {
        it ("gives the damager speed one") {
            val sprintDrain = SprintDrain(mockk())

            val effect = PotionEffect(PotionEffectType.SPEED, 5 * 20, 0)

            val damagerMock = mockk<Player> {
                every { addPotionEffect(effect, true) } returns true
            }

            val data = mockk<DamageEventEnchantData> {
                every { level } returns 1
                every { damager } returns damagerMock
            }

            sprintDrain.execute(data)

            verify { damagerMock.addPotionEffect(effect, true) }
        }

        describe("level 2 or greater") {
            it("gives the damager speed and gives the damagee slowness") {

            }
        }
    }
})
